### PR TITLE
BLD: Clean up build for complex

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by mattip** [comment]
Do you have a plan for the functions we blocklist for MSVC and for glibc<18? See https://github.com/numpy/numpy/blob/adf9794a1f699c7d9a8b9fde4cf3c3c4459aad35/numpy/core/src/common/npy_config.h#L21

**Comment by lysnikolaou** [comment]
I vaguely recalled that there was such a blocklist, but after all of CI passed locally on my machine (except for `npy_cpow`) I decided to open a PR to see what the situation is like on the rest of the supported platforms.

From here on, my plan's to test all of the currently block-listed functions. If there's a documented reason for blocklisting, add a test case to test it, in case there's none. If the tests (existing or new) fail, add it back our own implementation of the complex function back to the complex compat layer. Does that sound good?

**Comment by mattip** [comment]
OK. Maybe compare and contrast with some of the last changes made to the blocklist. Here are the [PRs I did previously](https://github.com/numpy/numpy/pulls?q=is%3Apr+author%3Amattip+is%3Aclosed+npymath)

**Comment by lysnikolaou** [comment]
@mattip @rgommers Unfortunately this is all we can remove without CI starting to fail. I also cleaned up the build process around complex stuff. Let me know if there's something I might have missed.

**Review by rgommers — COMMENTED** [review]
Thanks @lysnikolaou. I like all the build system cleanups, which is the bulk of this PR.

I'm less sure about the removals from the blocklist, because it is not really possible to tell if those things are okay now as far as I can tell. Do we actually need those changes, or can we just leave them out? We do have CI jobs that has build coverage for them (from gh-9567, `hypot`, `cabs`, `carg`), because we have a job with 32-bit Python + MSVC 2019. But it's unclear to me that we actually have a test for the precision word bug from gh-9567. The bug report mentions an issue that shows up with `gfortran`, so it's not covered by the NumPy Windows CI. If we could get a confirmation somehow that the bugs were fixed in MSVC (e.g., by asking the MS folks), that would be more comfortable.

File: numpy/core/feature_detection_cmath.h [hunk 1]
```diff
@@ -31,6 +31,8 @@ cfloat ctanhf(cfloat);
 float crealf(cfloat);
 float cimagf(cfloat);
 cfloat conjf(cfloat);
+float cabsf(cfloat);
+float cargf(cfloat);
```
Comment (rgommers) [inline numpy/core/feature_detection_cmath.h:35]: The changes in this file were probably intended to be reverted too, or did you leave them in on purpose?

Comment (lysnikolaou) [reply numpy/core/feature_detection_cmath.h:35]: Correct, reverting now.

**Review by rgommers — APPROVED** [review]
Thanks @lysnikolaou! In it goes.

**Pull request merged**